pkg/cmd/roachtest/tests/mvcc_gc.go

```diff
@@ -0,0 +1,334 @@
+
+package tests
+
+import (
+	"context"
+	gosql "database/sql"
+	"fmt"
+	"math"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/option"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/registry"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/roachprod/install"
+	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
+	"github.com/cockroachdb/cockroach/pkg/util/retry"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+	"github.com/cockroachdb/errors"
+)
+
+func registerMVCCGC(r registry.Registry) {
+	r.Add(registry.TestSpec{
+		Name:    "mvcc_gc",
+		Owner:   registry.OwnerKV,
+		Timeout: 15 * time.Minute,
+		Cluster: r.MakeClusterSpec(3),
+		Run:     runMVCCGC,
+	})
+}
+
+/*
+Test verifies that GC respects range tombstones and garbage collects range
+tombstones and that mvcc stats are correctly updated.
+
+Test sequence:
+- create kv test data schema
+- pre-split with 3 min expiration to allow merging to proceed in parallel after
+  initial data is loaded
+- load test data using KV workload
+- delete all data in kv table using range tombstones
+  (deletion is done using cli commands that sends kv requests to cluster)
+- sets low GC ttl to make all data eligible for deletion after 2 min
+- write more data over the existing one
+- verify that all garbage and range keys were removed by GC within 5 min
+*/
+
+func runMVCCGC(ctx context.Context, t test.Test, c cluster.Cluster) {
+	c.Put(ctx, t.Cockroach(), "./cockroach")
+	s := install.MakeClusterSettings()
+	s.Env = append(s.Env, "COCKROACH_SCAN_INTERVAL=30s")
+	c.Start(ctx, t.L(), option.DefaultStartOpts(), s)
+
+	conn := c.Conn(ctx, t.L(), 1)
+	defer conn.Close()
+
+	execSQLOrFail := func(statement string, args ...interface{}) {
+		if _, err := conn.ExecContext(ctx, statement, args...); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	setClusterSetting := func(name string, value interface{}) {
+		execSQLOrFail(fmt.Sprintf(`SET CLUSTER SETTING %s = $1`, name), value)
+	}
+
+	setClusterSetting("storage.mvcc.range_tombstones.enabled", true)
+	setClusterSetting("kv.protectedts.poll_interval", "5s")
+	setClusterSetting("kv.mvcc_gc.queue_interval", "0s")
+
+	m := c.NewMonitor(ctx)
+	m.Go(func(ctx context.Context) error {
+		c.Run(ctx, c.Node(1), "./cockroach", "workload", "init", "kv", "--cycle-length", "10000")
+
+		splitKvIntoRanges(t, conn, 100, 180*time.Second, "kv", "kv")
+
+		c.Run(ctx, c.Node(1), "./cockroach", "workload", "run", "kv", "--cycle-length", "10000",
+			"--max-block-bytes", "2048", "--min-block-bytes", "2048", "--duration", "2m",
+			"--read-percent", "0")
+
+		t.L().Printf("Finished prepping GC data")
+
+		if err := retry.WithMaxAttempts(ctx, retry.Options{}, 3, func() error {
+			return deleteAllTableDataWithTombstones(ctx, t, c, conn, "kv", "kv")
```

Comment:
Let's write several partial range tombstones across random spans of the keyspace, and do a few cycles of point writes and range deletions to get overlapping range tombstones.

We could consider resolving #86654 while we're at it, and lift the MVCC range tombstone writes into `workload`.

Reply:
It is pretty tricky to lift that to workload. If workload will generate a mix I'm not sure how we could assert stats in a meaningful way.
What I did is added a bunch of overlapping ranges that are guaranteed to cover all table space. With such covering we can assert that there's no live data and subsequently when GC cycle passes that there's no range tombstones as overlaying data has no range tombstones.

Reply:
> I'm not sure how we could assert stats in a meaningful way.

The only assertions we care about here are:

* The consistency check passes, i.e. the evaluated stats match the recomputed stats.
* GCBytesAge is 0.
* RangeKeyCount is 0.

That will confirm that all eligible data was GCed, and that GC updated stats correctly. We can do this regardless of what the workload wrote, we don't have to know anything about it.